korangar/src/loaders/animation/mod.rs

```diff
@@ -0,0 +1,474 @@
+            // This code resize the sprites in the same size and same offset
+            // Find the max_width and the min and max offset
+            let mut max_width = 0;
+            let mut max_height = 0;
+            let mut min_offset_x = i32::MAX;
+            let mut min_offset_y = i32::MAX;
+            let mut max_offset_x = 0;
+            let mut max_offset_y = 0;
+            frames.iter_mut().for_each(|frame| {
+                max_width = max(max_width, frame.size.x);
+                max_height = max(max_height, frame.size.y);
+                min_offset_x = min(min_offset_x, frame.offset.x);
+                min_offset_y = min(min_offset_y, frame.offset.y);
+                max_offset_x = max(max_offset_x, frame.offset.x);
+                max_offset_y = max(max_offset_y, frame.offset.y);
+            });
+            // The constant 2 is to avoid miscalculation by 1.
+            max_width += max_offset_x - min_offset_x + 2;
+            max_height += max_offset_y - min_offset_y + 2;
+
+            // Shift every frame by  max_offset_x - min_offset_x + 2
+            // As the bottom part and left part will not be in the offset.
+            frames.iter_mut().for_each(|frame| {
+                let new_width = max_width;
+                let new_height = max_height;
+                for framepart in frame.frameparts.iter_mut() {
+                    framepart.offset.x += max_offset_x - min_offset_x + 2;
+                    framepart.offset.y += max_offset_y - min_offset_y + 2;
+                }
+                frame.offset.x = min_offset_x;
+                frame.offset.y = min_offset_y;
+                frame.offset.x += max_offset_x - min_offset_x + 2;
+                frame.offset.y += max_offset_y - min_offset_y + 2;
+                frame.size = Vector2::new(new_width, new_height);
+                frame.remove_offset.x = max_offset_x - min_offset_x + 2;
+                frame.remove_offset.y = max_offset_y - min_offset_y + 2;
+            });
+            animations.push(Animation { frames });
+        }
+        let animation_data = AnimationData {
+            delays: (&vec[0]).actions.delays.clone(),
+            animation_pair: vec,
+            animations,
+            entity_type,
+        };
+        let hash = match entity_type {
+            EntityType::Player => format!("0_{}_{}", entity_filename[0], entity_filename[1]),
+            EntityType::Monster => format!("1_{}", entity_filename[0]),
+            EntityType::Npc => format!("2_{}", entity_filename[0]),
+            _ => format!("3"),
+        };
+        self.cache.insert(hash.clone(), animation_data.clone());
+        Ok(animation_data)
+    }
+
+    pub fn get(
+        &mut self,
+        sprite_loader: &mut SpriteLoader,
+        action_loader: &mut ActionLoader,
+        entity_filename: Vec<String>,
+        entity_type: EntityType,
+    ) -> Result<AnimationData, LoadError> {
+        let hash = match entity_type {

```

Comment:
I think it's better to create a enum here, that we then can use as a hash instead (an enum that has fields for each type). Encoded strings are very brittle.

Reply:
I am going to change the hash function later, for now we use the inefficient string, as the commit is to big. The npc and monster is have id with number 30k or less, for player we can offset by 30k and create a number for each player.

Reply:
The change could be as simple as:

```
#[derive(Clone, Debug, Hash)]
enum AnimationKey {
    Player { body: String, head: String },
    Monster(String),
    Npc(String),
}
```

Reply:
I changed to use number instead, see if it is okay.

Reply:
Okay, I think I know what you aimed for here. But since you are using the EntityId now, you can't really cache the values anymore, since EntityID's are not guranteed to be unique over the lifetime of the game session!

That would mean, as soon as an entity despawns, you have to remove it's animation from the cache, or else we could have  a wrong cache hit.

If we want to cache the animations,  then we need something that makes them identifiable for that they are. For example: A sprite is identifiable using it's path inside the game files. I think your approach using the file names was already good, but only needed a little refinement. What would you need to re-load the same animation again? I thought the filenames was already a good fit.

Reply:
The hashmap is not using the entity id from the session, it is using the id linked with the sprite search in the Lua script. For now npc and monster is ok in cache, for player need to change later, i only passing the job id and sex for now and transform it in a number, later we can hash the parameters in other ways.

Reply:
Than please don't call it "entity_id". That's already a well established term!
If you know the ID's are unique (or make sure they are), then it's fine, but you can then also use an enum to make this better to use:

```
#[derive(Clone, Debug, Hash)]
enum AnimationKey {
    Player { job: u32, head: u32, is_female: bool },
    Monster(u32),
    Npc(u32),
}
```

Reply:
Fixed